<a href="https://colab.research.google.com/github/amsa10/Conv-segmentation/blob/main/task2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [4]:
import numpy as np

# Define the weights
weights = {
    'w11': 0.73, 'w12': -1.23,
    'w21': 1.18, 'w22': 0.44,
    'w13': 1.13, 'w23': 0.96,
    'w14': 0.46, 'w24': -0.76
}

# Define the activations of the last layer
u31 = 0.53
u32 = 0.41

# Sigmoid activation function
def sigmoid(x):
    return 1 / (1 + np.exp(-x))

# Calculate the activations of the hidden layer 2
z21 = weights['w11'] * u31 + weights['w12'] * u32
z22 = weights['w21'] * u31 + weights['w22'] * u32

u21 = sigmoid(z21)
u22 = sigmoid(z22)

# Calculate the activations of the hidden layer 1
# To get u11 and u12, we need the values of weights connecting the input to the first hidden layer
# Assume we have similar weights for the hidden layer 1
# Note: This is an example, you need to adjust according to actual weights and inputs

# For demonstration, let's assume we have hypothetical weights to compute u11 and u12
weights_hidden_layer_1 = {
    'w31': weights['w13'], 'w32': weights['w23'],
    'w41': weights['w14'], 'w42': weights['w24']
}

# Let's assume some input values (placeholders)
input1 = 1.0  # Placeholder value
input2 = 1.0  # Placeholder value

# Calculate the activations for the first hidden layer
z11 = weights_hidden_layer_1['w31'] * input1 + weights_hidden_layer_1['w32'] * input2
z12 = weights_hidden_layer_1['w41'] * input1 + weights_hidden_layer_1['w42'] * input2

u11 = sigmoid(z11)
u12 = sigmoid(z12)

# Print results
print(f'u21: {u21}')
print(f'u22: {u22}')
print(f'u11: {u11}')
print(f'u12: {u12}')




u21: 0.4706836639363989
u22: 0.6912137883600998
u11: 0.8899274256374969
u12: 0.425557483188341


In [6]:
import numpy as np


# Given weights
weights = {
    'w11': 0.73, 'w12': -1.23,
    'w21': 1.18, 'w22': 0.44,
    'w13': 1.13, 'w23': 0.96,
    'w14': 0.46, 'w24': -0.76
}

# Given activations
u31 = 0.53
u32 = 0.41

# Learning rate
learning_rate = 0.05

# Define sigmoid and its derivative
def sigmoid(x):
    return 1 / (1 + np.exp(-x))

def sigmoid_derivative(x):
    return sigmoid(x) * (1 - sigmoid(x))

# Assume some hypothetical target values and predictions
# For the sake of this example, let's use:
target = np.array([1, 0])  # Assuming class K2
output = np.array([u31, u32])  # These are the activations from the output layer

# Compute the error
error = target - output

# Compute the gradient of the loss w.r.t the output layer activations
output_grad = -error * sigmoid_derivative(output)

# Calculate gradients with respect to the weights
grad_w12 = output_grad[0] * u31
grad_w22 = output_grad[1] * u32

# Update the weights
weights['w12'] -= learning_rate * grad_w12
weights['w22'] -= learning_rate * grad_w22

# Output the updated weights
print(f"Updated weight w12: {weights['w12']}")
print(f"Updated weight w22: {weights['w22']}")



Updated weight w12: -1.2270950689891267
Updated weight w22: 0.4379846386714968


Updated weight w12: -1.2270950689891267:

Это значение указывает, что вес
𝑤
12
w
12
​
  немного изменился по сравнению с исходным значением. Он стал немного менее отрицательным.
Updated weight w22: 0.4379846386714968:

Это значение показывает, что вес
𝑤
22
w
22
​
  уменьшился, что делает его менее положительным.